### PR TITLE
fixes issues with super->sub-class auto-cast and handles MultiInputObj coercion

### PR DESCRIPTION
**Pull request by tclose**
## Types of changes

- Bug fix (non-breaking change which fixes an issue)

## Summary
* fixes bugs with super->sub-class auto-casting, including when the sub-classes are part of a union. If the super class is a super class of any of the union args then it passes the check.
* adds handling for MultiInputObj
* streamlines typing error messages

## Checklist

- [x] I have added tests to cover my changes (if necessary)

### CODE REVIEW

**Comment by tclose** [comment]
NB: the reference hash updates required in the unittests seem be related to changes to the file-hashing process. This could be due to a change in `fileformats` but I can't remember making any changes that would be related.

File: pydra/utils/tests/test_typing.py [hunk 1]
```diff
@@ -610,6 +652,18 @@ def test_contains_type_in_dict():
     )
 
 
+def test_any_union():
+    """Check that the superclass auto-cast matches if any of the union args match instead
+    of all"""
+    TypeParser(File, match_any_of_union=True).check_type(ty.Union[ty.List[File], Json])
```
[the remark this answers is not in the document]
Comment (tclose) [reply pydra/utils/tests/test_typing.py:658]: yes, that is right.

File: pydra/utils/tests/test_typing.py [hunk 2]
```diff
@@ -610,6 +652,18 @@ def test_contains_type_in_dict():
...
+
+
+def test_union_superclass_check_type():
+    """Check that the superclass auto-cast matches if any of the union args match instead
+    of all"""
+    TypeParser(ty.Union[ty.List[File], Json], superclass_auto_cast=True)(lz(File))
```
[the remark this answers is not in the document]
Comment (tclose) [reply pydra/utils/tests/test_typing.py:664]: yes

[the remark this answers is not in the document]
Comment (effigies) [reply pydra/utils/tests/test_typing.py:664]: Maybe a comment, or expansion of the docstring, to make that super clear? It took me a little reading to understand, so I figure it's likely to take time again later.

**Comment by tclose** [comment]
@effigies I have parametrized a number of test batches, there a couple more that could probably be done but would require a bit more work so I thought I could leave them for now.

@djarecka if you are busy are you happy to merge if @effigies has been through it? It just cleans up some corner cases that were missed in the original PR

File: pydra/utils/typing.py [hunk 3]
```diff
@@ -345,7 +368,26 @@ def coerce_obj(obj, type_):
...
+                    return [inner_type_parser.coerce(object_)]
+                except TypeError:
+                    add_exc_note(
+                        e,
+                        "Also failed to coerce to the arg-type of the MultiInputObj "
+                        f"({self.pattern[1][0]})",
+                    )
+                    raise e
+            else:
+                raise e
```
Comment (effigies) [inline pydra/utils/typing.py:390]: Another re-ordering of branching.

```suggestion
        except TypeError as e:
            # Special handling for MultiInputObjects (which are annoying)
            if not isinstance(self.pattern, tuple) or self.pattern[0] != MultiInputObj:
                raise e
            # Attempt to coerce the object into arg type of the MultiInputObj first,
            # and if that fails, try to coerce it into a list of the arg type
            inner_type_parser = copy(self)
            inner_type_parser.pattern = self.pattern[1][0]
            try:
                return [inner_type_parser.coerce(object_)]
            except TypeError:
                add_exc_note(
                    e,
                    "Also failed to coerce to the arg-type of the MultiInputObj "
                    f"({self.pattern[1][0]})",
                )
                raise e
```

**Review by effigies — APPROVED** [review]
It looks like `coerce()` and `check_type()` have a lot of duplicated code. Does not need to be addressed now, but this is inviting desynchronization and fixing bugs in one and not the other.

Otherwise LGTM.

**Pull request merged**